In [ ]:
#Importing libraries for data manipulation and linear algebra
import pandas as pd
import numpy as np

#Importing libraries for data visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

#Importing library train and test split
from sklearn.model_selection import train_test_split

#Importing library for train a model
from sklearn.ensemble import RandomForestClassifier

#Importing library for extract accuracy
from sklearn.metrics import accuracy_score

#Importing library for finding constant & quasi constant variance
from sklearn.feature_selection import VarianceThreshold

In [ ]:
#Loading the data
train = pd.read_csv('https://raw.githubusercontent.com/laxmimerit/Data-Files-for-Feature-Selection/master/santander-train.csv',
                   nrows = 1000)

In [ ]:
#Checking the data
train.head()

ID  var3  var15  ...  saldo_medio_var44_ult3          var38  TARGET
0   1     2     23  ...                     0.0   39205.170000       0
1   3     2     34  ...                     0.0   49278.030000       0
2   4     2     23  ...                     0.0   67333.770000       0
3   8     2     37  ...                     0.0   64007.970000       0
4  10     2     39  ...                     0.0  117310.979016       0

[5 rows x 371 columns]

In [ ]:
#Checking the shape of the data
train.shape

(1000, 371)

Data have 1000 observations and 371 features

In [ ]:
#Splitting data into X & y column wise
X = train.drop(columns = 'TARGET', axis=1)
y = train['TARGET']

#Checking the shape
X.shape, y.shape

((1000, 370), (1000,))

In [ ]:
#Splitting X & y in train and test(80%-20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42, stratify = y)

#Checking the shape of the data
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((800, 370), (800,), (200, 370), (200,))

#Constant features removal

In [ ]:
#Creating constant filter
constant_filter = VarianceThreshold(threshold=0)

#Constant filter fit on the train data
constant_filter.fit(X_train)

VarianceThreshold(threshold=0)

In [ ]:
#Checking the result
constant_filter.get_support()

array([ True,  True,  True,  True,  True,  True,  True,  True, False,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False, False,  True,  True,  True,  True,  True, False,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False, False, False, False,  True,  True,
        True,  True,  True,  True,  True,  True,  True, False, False,
       False, False,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False, False,  True,  True,  True,
        True,  True, False, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

Above 'True' indicate non constant features and 'False' indicate constant features

In [ ]:
#Checking total non-constant features
constant_filter.get_support().sum()

259

In [ ]:
#Checking which are the features are constant
constant_features = X_train.columns[~constant_filter.get_support()]
constant_features

Index(['imp_op_var40_efect_ult1', 'ind_var2_0', 'ind_var2',
       'ind_var13_medio_0', 'ind_var13_medio', 'ind_var18_0', 'ind_var18',
       'ind_var27_0', 'ind_var28_0', 'ind_var28',
       ...
       'saldo_medio_var13_medio_hace3', 'saldo_medio_var13_medio_ult1',
       'saldo_medio_var13_medio_ult3', 'saldo_medio_var17_hace3',
       'saldo_medio_var29_hace2', 'saldo_medio_var29_hace3',
       'saldo_medio_var33_hace2', 'saldo_medio_var33_hace3',
       'saldo_medio_var33_ult1', 'saldo_medio_var33_ult3'],
      dtype='object', length=111)

In [ ]:
#Checking which are the features are constant
non_constant_features = X_train.columns[constant_filter.get_support()]
non_constant_features

Index(['ID', 'var3', 'var15', 'imp_ent_var16_ult1', 'imp_op_var39_comer_ult1',
       'imp_op_var39_comer_ult3', 'imp_op_var40_comer_ult1',
       'imp_op_var40_comer_ult3', 'imp_op_var40_efect_ult3',
       'imp_op_var40_ult1',
       ...
       'saldo_medio_var17_hace2', 'saldo_medio_var17_ult1',
       'saldo_medio_var17_ult3', 'saldo_medio_var29_ult1',
       'saldo_medio_var29_ult3', 'saldo_medio_var44_hace2',
       'saldo_medio_var44_hace3', 'saldo_medio_var44_ult1',
       'saldo_medio_var44_ult3', 'var38'],
      dtype='object', length=259)

In [ ]:
#Checking the count of constant and non constant features
print('Constant features:',len(constant_features))
print('Non-Constant features:',len(non_constant_features))

Constant features: 111
Non-Constant features: 259


In [ ]:
#Viewing the constant features values in the data
for i in X_train[constant_features].columns:
  print(i,':')
  print(X_train[i].value_counts())
  print('---------------------------------------')

imp_op_var40_efect_ult1 :
0    800
Name: imp_op_var40_efect_ult1, dtype: int64
---------------------------------------
ind_var2_0 :
0    800
Name: ind_var2_0, dtype: int64
---------------------------------------
ind_var2 :
0    800
Name: ind_var2, dtype: int64
---------------------------------------
ind_var13_medio_0 :
0    800
Name: ind_var13_medio_0, dtype: int64
---------------------------------------
ind_var13_medio :
0    800
Name: ind_var13_medio, dtype: int64
---------------------------------------
ind_var18_0 :
0    800
Name: ind_var18_0, dtype: int64
---------------------------------------
ind_var18 :
0    800
Name: ind_var18, dtype: int64
---------------------------------------
ind_var27_0 :
0    800
Name: ind_var27_0, dtype: int64
---------------------------------------
ind_var28_0 :
0    800
Name: ind_var28_0, dtype: int64
---------------------------------------
ind_var28 :
0    800
Name: ind_var28, dtype: int64
---------------------------------------
ind_var27 :
0    800
N

As we found that 111 features have 0 values

In [ ]:
#Removing constant features from the data
X_train_filter = constant_filter.transform(X_train)
X_test_filter = constant_filter.transform(X_test)

#Checking the shape of the data
X_train_filter.shape, X_test_filter.shape

((800, 259), (200, 259))

We have removed 111 constant variables from the data and now We have 259 variables

In [ ]:
#Converting numpy array form of filter data into dataframe
X_train_filter = pd.DataFrame(X_train_filter, columns=non_constant_features)
X_test_filter = pd.DataFrame(X_test_filter, columns=non_constant_features)

# Quasi constant features removal

In [ ]:
#Creating quasi constant filter
quasi_constant_filter = VarianceThreshold(threshold=0.01)

#Fit the quasi constant filter on the X_train_filter data
quasi_constant_filter.fit(X_train_filter)

VarianceThreshold(threshold=0.01)

In [ ]:
#Checking the total of non quasi constant features
quasi_constant_filter.get_support().sum()

234

In [ ]:
#Checking the result of quasi constant filter
quasi_constant_filter.get_support()

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True, False,  True,  True, False, False,  True,  True,
        True,  True,  True,  True,  True, False, False,  True,  True,
       False, False, False, False, False, False,  True,  True,  True,
        True,  True,  True,  True,  True, False, False, False,  True,
       False, False, False, False, False,  True,  True,  True,  True,
        True, False,  True, False, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

Above 'True' are a non quasi constant features and 'False' are a quasi constant features

In [ ]:
#Checking Which variables are a quasi constant features in the data
quasi_constant_features = X_train_filter.columns[~quasi_constant_filter.get_support()]
quasi_constant_features

Index(['ind_var1', 'ind_var6_0', 'ind_var6', 'ind_var13_largo_0',
       'ind_var13_largo', 'ind_var14', 'ind_var17_0', 'ind_var17', 'ind_var19',
       'ind_var20_0', 'ind_var20', 'ind_var29_0', 'ind_var29', 'ind_var30_0',
       'ind_var31_0', 'ind_var31', 'ind_var32_cte', 'ind_var32_0', 'ind_var32',
       'ind_var40', 'ind_var39', 'ind_var44_0', 'ind_var44',
       'ind_var7_recib_ult1', 'num_meses_var29_ult3'],
      dtype='object')

In [ ]:
#Checking Which variables are a non-quasi constant features in the data
non_quasi_constant_features = X_train_filter.columns[quasi_constant_filter.get_support()]
non_quasi_constant_features

Index(['ID', 'var3', 'var15', 'imp_ent_var16_ult1', 'imp_op_var39_comer_ult1',
       'imp_op_var39_comer_ult3', 'imp_op_var40_comer_ult1',
       'imp_op_var40_comer_ult3', 'imp_op_var40_efect_ult3',
       'imp_op_var40_ult1',
       ...
       'saldo_medio_var17_hace2', 'saldo_medio_var17_ult1',
       'saldo_medio_var17_ult3', 'saldo_medio_var29_ult1',
       'saldo_medio_var29_ult3', 'saldo_medio_var44_hace2',
       'saldo_medio_var44_hace3', 'saldo_medio_var44_ult1',
       'saldo_medio_var44_ult3', 'var38'],
      dtype='object', length=234)

In [ ]:
#Checking count of quasi constant and non quasi constant features
print('Quasi Constant Features:', len(quasi_constant_features))
print('Non-Quasi Constant Features:', len(non_quasi_constant_features))

Quasi Constant Features: 25
Non-Quasi Constant Features: 234


In [ ]:
#Checking quasi constant features and its count
for i in X_train_filter[quasi_constant_features].columns:
  print(i,':')
  print(X_train_filter[i].value_counts())
  print('--------------------------------------')

ind_var1 :
0.0    794
1.0      6
Name: ind_var1, dtype: int64
--------------------------------------
ind_var6_0 :
0.0    799
1.0      1
Name: ind_var6_0, dtype: int64
--------------------------------------
ind_var6 :
0.0    799
1.0      1
Name: ind_var6, dtype: int64
--------------------------------------
ind_var13_largo_0 :
0.0    793
1.0      7
Name: ind_var13_largo_0, dtype: int64
--------------------------------------
ind_var13_largo :
0.0    793
1.0      7
Name: ind_var13_largo, dtype: int64
--------------------------------------
ind_var14 :
0.0    798
1.0      2
Name: ind_var14, dtype: int64
--------------------------------------
ind_var17_0 :
0.0    797
1.0      3
Name: ind_var17_0, dtype: int64
--------------------------------------
ind_var17 :
0.0    797
1.0      3
Name: ind_var17, dtype: int64
--------------------------------------
ind_var19 :
0.0    797
1.0      3
Name: ind_var19, dtype: int64
--------------------------------------
ind_var20_0 :
0.0    794
1.0      6
Name: i

In [ ]:
#Removing quasi constant from the data
X_train_quasi_filter = quasi_constant_filter.transform(X_train_filter)
X_test_quasi_filter = quasi_constant_filter.transform(X_test_filter)

#Checking the shape of the data
X_train_quasi_filter.shape, X_test_quasi_filter.shape

((800, 234), (200, 234))

In [ ]:
#Converting numpy array of data into dataframe
X_train_quasi_filter = pd.DataFrame(X_train_quasi_filter, columns=non_quasi_constant_features)
X_test_quasi_filter = pd.DataFrame(X_test_quasi_filter, columns=non_quasi_constant_features)

#Removing Duplicate Variables

In [ ]:
#Transposing the data
X_train_T = X_train_quasi_filter.T
X_test_T = X_test_quasi_filter.T

In [ ]:
#Checking the type of the data
type(X_train_T)

pandas.core.frame.DataFrame

In [ ]:
#Checking shape of the data
X_train_T.shape, X_test_T.shape

((234, 800), (234, 200))

In [ ]:
#Extracting count of duplicate rows
X_train_T.duplicated().sum()

33

In [ ]:
#Checking the result of duplicate function
X_train_T.duplicated()

ID                         False
var3                       False
var15                      False
imp_ent_var16_ult1         False
imp_op_var39_comer_ult1    False
                           ...  
saldo_medio_var44_hace2    False
saldo_medio_var44_hace3    False
saldo_medio_var44_ult1     False
saldo_medio_var44_ult3     False
var38                      False
Length: 234, dtype: bool

In [ ]:
#Transpose the rows into columns which was previously transposed columns into rows
X_train_NT = X_train_T.T
X_test_NT = X_test_T.T

In [ ]:
#Checking data
X_train_NT

ID  var3  ...  saldo_medio_var44_ult3          var38
0    1210.0   2.0  ...                    0.00  117310.979016
1    1917.0   2.0  ...                    0.00   39151.530000
2     609.0   2.0  ...                10601.67  117310.979016
3    1843.0   2.0  ...                    0.00  110254.350000
4    1951.0   2.0  ...                    0.00   95919.960000
..      ...   ...  ...                     ...            ...
795  1007.0   2.0  ...                    0.00   98790.540000
796  1126.0   2.0  ...                    0.00  117310.979016
797   737.0   2.0  ...                    0.00  268504.080000
798  1411.0   2.0  ...                    0.00   54614.130000
799  1271.0   2.0  ...                    0.00   41779.980000

[800 rows x 234 columns]

In [ ]:
#Extracting duplicate features from the data
duplicate_features = X_train_NT.columns[X_train_T.duplicated()]
duplicate_features

Index(['imp_op_var39_efect_ult1', 'ind_var26', 'ind_var25', 'ind_var37',
       'ind_var40_0', 'num_var6', 'num_var13_largo', 'num_var17', 'num_var26',
       'num_var25', 'num_var29_0', 'num_var29', 'num_var31', 'num_var32',
       'num_var37', 'num_var40_0', 'num_var40', 'num_var39', 'num_var44',
       'saldo_var29', 'saldo_var40', 'delta_imp_trasp_var17_in_1y3',
       'delta_num_aport_var13_1y3', 'delta_num_aport_var17_1y3',
       'delta_num_compra_var44_1y3', 'delta_num_reemb_var13_1y3',
       'delta_num_trasp_var17_in_1y3', 'delta_num_venta_var44_1y3',
       'imp_trasp_var17_in_ult1', 'num_op_var39_efect_ult1',
       'num_trasp_var17_in_ult1', 'num_venta_var44_ult1',
       'saldo_medio_var29_ult3'],
      dtype='object')

In [ ]:
#Extracting non-duplicate features from the data
non_duplicate_features = X_train_NT.columns[~X_train_T.duplicated()]
non_duplicate_features

Index(['ID', 'var3', 'var15', 'imp_ent_var16_ult1', 'imp_op_var39_comer_ult1',
       'imp_op_var39_comer_ult3', 'imp_op_var40_comer_ult1',
       'imp_op_var40_comer_ult3', 'imp_op_var40_efect_ult3',
       'imp_op_var40_ult1',
       ...
       'saldo_medio_var13_largo_ult3', 'saldo_medio_var17_hace2',
       'saldo_medio_var17_ult1', 'saldo_medio_var17_ult3',
       'saldo_medio_var29_ult1', 'saldo_medio_var44_hace2',
       'saldo_medio_var44_hace3', 'saldo_medio_var44_ult1',
       'saldo_medio_var44_ult3', 'var38'],
      dtype='object', length=201)

In [ ]:
#Checking count of duplicate and non duplicate features
print('Duplicate features:', len(duplicate_features))
print('Non Duplicate features:',len(non_duplicate_features))

Duplicate features: 33
Non Duplicate features: 201


In [ ]:
#Removing duplicate features from the data
X_train_unique = X_train_NT[non_duplicate_features]
X_test_unique = X_test_NT[non_duplicate_features]

#Checking the shape of the data
X_train_unique.shape, X_test_unique.shape

((800, 201), (200, 201))

In [ ]:
#Creating function for train a model on data
def classifier(model, X_train, X_test, y_train, y_test):
  model.fit(X_train, y_train)
  y_pred = model.predict(X_test)
  accuracy = accuracy_score(y_test, y_pred)
  return accuracy

In [ ]:
%%time
#Train a model, predict on the model and extract accuracy
rf = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
print('Accuracy on 201 unique variables 201:',classifier(rf, X_train_unique, X_test_unique, y_train, y_test))

Accuracy on 201 unique variables 201: 0.96
CPU times: user 364 ms, sys: 29.1 ms, total: 393 ms
Wall time: 454 ms


In [ ]:
%%time
#Train a model, predict on the model and extract accuracy
rf = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
print('Accuracy on 370 original variables:',classifier(rf, X_train, X_test, y_train, y_test))

Accuracy on 370 original variables: 0.96
CPU times: user 365 ms, sys: 39.1 ms, total: 404 ms
Wall time: 463 ms


We have removed constant features, quasi constant features and duplicate features after which We have generated a model where We got 96% model accuracy and it took around 454 millisecond whereas We have generated a model on original data where We got 96% model accuracy and it took around 463 millisecond.

Conclusion:
As the accuracy comparision no improvement in accuracy however model took lesser time as compare to model on original data. We have used basic feature selection methods which helped us to reduce time computation. 